packages/eslint-plugin/docs/rules/no-misused-promises.mdx

```diff
@@ -100,6 +100,8 @@ The following options are supported:
 
 - `arguments`: Disables checking an asynchronous function passed as argument where the parameter type expects a function that returns `void`
 - `attributes`: Disables checking an asynchronous function passed as a JSX attribute expected to be a function that returns `void`
+- `heritageTypes`: Disables checking an asynchronous method in a type whose extended/implemented heritage type expects that method to return `void`
```

Comment:
[Naming] I don't love how this is the only option that mentions _"Types"_ in its name. It's redundant. But just _"heritage"_ on its own seems imprecise... hmm. Maybe, `heritageSignatures`? That's kind of wordy but 🤷 I can't think of anything more succinct. What do you think?

Reply:
Ah I didn't necessarily see it as redundant since I'm using "heritage types" to refer specifically to the super-types aka heritage/base types that are extended/implemented. I.e. I meant "this option checks a subtype's async methods against its heritage types". But I'm also happy to go with another name, and maybe `heritageTypes` is inconsistent anyway if the other options refer to what we're checking (not what we're checking *against*).

Maybe `subtypeMethods`?

E.g.

- `subtypeMethods`: Disables checking an asynchronous method in a subtype whose base type expects that method to return `void`

For reference here's the [TS Handbook section on Class Heritage](https://www.typescriptlang.org/docs/handbook/2/classes.html#class-heritage) in case that helps spark any better alternatives for you (`subtypeMethods` is the most precise I can think of right now).

Re: `heritageSignatures` I wonder if "signatures" is more specific than we intend, depending on what we mean by signature? And maybe it's better to focus on the `subtype` since that's what we're checking

Reply:
Nice yeah I like it, although do you happen to know if *derived* also applies when a subtype implements an interface?

Reply:
Ah, you're right, _derived_ just refers to classes _deriving_ methods from their base. Good call! 

How about `inheritedMethods`?

Reply:
Ah okay thanks for confirming that. Hmm wouldn't `inheritedMethods` have the same issue as `derived`? A class doesn't inherit an interface method, it just implements it, as I understand it?

Reply:
Whoops, sorry, I missed this till now!

My gut instinct tells me that difficulties in clearly naming something generally come from that thing being too broad. Which makes sense here! It's one option trying to do two things: `extends` and `inherits`. The two are similar on the surface but are actually two different things.

 How about splitting it into two options? Maybe, `extensions` and `implementations`?

Reply:
No prob! That's an interesting point. You're right we are checking two different things (extended types vs implemented types). Although under the surface we are actually just doing one thing, i.e. a single pass through `tsNode.heritageClauses` and running the same code on each heritage clause regardless of its type.

So if we wanted to split it into two options, we'd need to change up that singular code to, instead, first filter `heritageClauses` based on the user's choices (only check `extends` clauses, only check `implements` clauses, or check both); and then still run the same code on the filtered clauses. 

I wonder if this would provide extra value, or potentially just be more confusing for a user vs having a single option to check heritage methods. I think it's likely a similar intention either way: a subtype should be interchangable with its heritage type, e.g. `Dog.bark` should have the same return value as `Animal.bark` whether `Dog extends Animal` or `Dog implements Animal`. And the split options + filtering logic would add extra complexity to the rule implementation (especially e.g. where we [map node types to handlers based on options](https://github.com/typescript-eslint/typescript-eslint/pull/8765/files#diff-dd469c96cb4a7fd7e266d9957762126ca0e669b65eb00173edcf40d1dcdced6fR154-R179)). Let me know you can think of a use case here though, for wanting to check one but not the other.

At the moment I still lean towards `heritageMethods` being the most descriptive and simplest / least confusing. And a good thing is that the word "Heritage" does come straight from both the [TS docs](https://www.typescriptlang.org/docs/handbook/2/classes.html#class-heritage) and TS source code, and is used to cover both `extends` and `implements` clauses there. But if you still think splitting it into two options is more ideal let me know and I can make those changes. Otherwise if you think my logic here makes sense I can just change the name to `heritageMethods` and we'll go with that. Let me know either way, happy to do whatever you think is best :)

Reply:
cc @typescript-eslint/triage-team - let's see if anybody else on the team has input.

Reply:
My 2c:

- I prefer sticking to one option, unless someone complains with a concrete use case of why they'd want to have them specified separately

- I think we need a quick code example in the docs in order to make clear what this option means.

- I did not know the specific meaning of the word "heritage" in this context, and I'm still intimidated by it 😅 . I think it would be acceptable, but not preferable to use, provided there is explanation and example. I would prefer we use the word "inherited" instead. I personally like "inheritedMethodSignatures". or "inheritedMethods". Not super opinionated on the naming though.

Reply:
Hey @kirkwaiblinger  thanks for the thoughts :)

> * I think we need a quick code example in the docs in order to make clear what this option means.

I'd actually already added [an incorrect example](https://github.com/typescript-eslint/typescript-eslint/pull/8765/files#diff-64a7137a0db4419afd7a036a5fa31d285bb7891363e29db109766e41e28141f1R184-R192) and [a correct example](https://github.com/typescript-eslint/typescript-eslint/pull/8765/files#diff-64a7137a0db4419afd7a036a5fa31d285bb7891363e29db109766e41e28141f1R236-R244) to the `checksVoidReturn: true` examples section of the doc:

```ts
interface MySyncInterface {
  setThing(): void;
}
class MyClass implements MySyncInterface {
  async setThing(): Promise<void> {
    this.thing = await fetchThing();
  }
}

...

interface MyAsyncInterface {
  setThing(): Promise<void>;
}
class MyClass implements MyAsyncInterface {
  async setThing(): Promise<void> {
    this.thing = await fetchThing();
  }
}
```

 Do you think those are sufficient? Or is there anything that would help the clarity here?

> I did not know the specific meaning of the word "heritage" in this context, and I'm still intimidated by it 😅 . I think it would be acceptable, but not preferable to use, provided there is explanation and example. I would prefer we use the word "inherited" instead. I personally like "inheritedMethodSignatures". or "inheritedMethods". Not super opinionated on the naming though.

Haha, yeah that's fair enough. I think potentially the upside of `heritageMethods` is that it's more accurate since it refers to both `extends` and `implements` (vs `inheritedMethods` technically doesn't apply to e.g. `MyClass implements MySyncInterface`). But if it would be less intimidating/clearer to use `inheritedMethods` I could see that being worth it, and we could clarify in the option's description that it applies to both implemented and inherited/overridden methods. I can also see an argument that `MyClass` does inherit method signatures from `MySyncInterface`, so maybe `inheritedMethodSignatures` is fully accurate (albeit longer and maybe still ambiguous - is it inherited signatures, or signatures of inherited methods?).

**TLDR** I can see any of `heritageMethods`, `inheritedMethods`, `inheritedMethodSignatures` being good enough with small tradeoffs, and probably lean towards `heritageMethods` or `inheritedMethods` for brevity

Reply:
It seems "inheritedMethods" is the unique element of the intersection of our preferences 🚀

Reply:
I still wish we could find a single-word name 😞 but this compromise is fine with me.